File: jenjin-core/src/test/java/com/binarymonks/jj/pools/PoolsTest.java
package com.binarymonks.jj.pools;

import org.junit.Before;
import org.junit.Test;
import org.junit.Assert;

public class PoolsTest {

    Pools test_obj;

    @Before
    public void setup() {
        test_obj = new Pools();
    }

    @Test(expected = Pools.NoPoolManagerException.class)
    public void getNew_WhenNoPoolManager_NotPoolable() {
        this.test_obj.nuw(ThingToPool.class);
    }

    @Test()
    public void getNew_Poolable() {
        PoolableThing newThing = this.test_obj.nuw(PoolableThing.class);
        Assert.assertNotNull(newThing);
    }

    @Test(expected = Pools.NoPoolManagerException.class)
    public void recycle_WhenNoPoolManager_NotPoolable() {
        this.test_obj.recycle(new ThingToPool());
    }

    @Test()
    public void getNew_NotPoolable_WithPoolManagerRegistered() {
        ThingPoolManager poolManager = new ThingPoolManager(3, 5);
        this.test_obj.registerManager(poolManager, ThingToPool.class);

        ThingToPool actual = this.test_obj.nuw(ThingToPool.class);

        Assert.assertEquals(actual.someValue, poolManager.initialise_value);
    }

    @Test()
    public void getNew_NotPoolable_AfterRecycling() {
        ThingPoolManager poolManager = new ThingPoolManager(3, 5);
        this.test_obj.registerManager(poolManager, ThingToPool.class);

        ThingToPool actual = this.test_obj.nuw(ThingToPool.class);

        Assert.assertEquals(actual.someValue, poolManager.initialise_value);

        this.test_obj.recycle(actual);

        actual = this.test_obj.nuw(ThingToPool.class);

        Assert.assertEquals(actual.someValue, poolManager.reset_value);
    }

    @Test()
    public void getNew_Poolable_AfterRecycling() {

        PoolableThing actual = this.test_obj.nuw(PoolableThing.class);


        this.test_obj.recycle(actual);

        PoolableThing actual_v2 = this.test_obj.nuw(PoolableThing.class);

        Assert.assertSame(actual, actual_v2);
        Assert.assertTrue(actual.reset);

    }

    public static class ThingPoolManager implements PoolManager<ThingToPool> {
        int initialise_value;
        int reset_value;

        public ThingPoolManager(int initialize_value, int reset_value) {
            this.initialise_value = initialize_value;
            this.reset_value = reset_value;
        }

        @Override
        public void reset(ThingToPool thingToPool) {
            thingToPool.someValue = reset_value;
        }

        @Override
        public ThingToPool create_new() {
            ThingToPool newObj = new ThingToPool();
            newObj.someValue = initialise_value;
            return newObj;
        }

        @Override
        public void dispose(ThingToPool thingToPool) {

        }
    }

    public static class ThingToPool {
        int someValue;
    }

    public static class PoolableThing implements Poolable {

        boolean reset = false;

        @Override
        public void reset() {
            reset = true;
        }

    }

}

File: jenjin-core/src/test/kotlin/com/binarymonks/jj/core/pools/PoolsTest.kt
package com.binarymonks.jj.core.pools

import org.junit.Assert
import org.junit.Before
import org.junit.Test

class PoolsTest {

    lateinit var testObj: Pools

    @Before
    fun setup() {
        testObj = Pools()
    }

    @Test(expected = Pools.NoPoolManagerException::class)
    fun getNew_WhenNoPoolManager_NotPoolable() {
        testObj.nuw(SceneToPool::class.java)
    }

    @Test
    fun getNew_Poolable() {
        val newScene = testObj.nuw(PoolableScene::class.java)
        Assert.assertNotNull(newScene)
    }

    @Test(expected = Pools.NoPoolManagerException::class)
    fun recycle_WhenNoPoolManager_NotPoolable() {
        testObj.recycle(SceneToPool())
    }

    @Test
    fun getNew_NotPoolable_WithPoolManagerRegistered() {
        val poolManager = ScenePoolManager(3, 5)
        testObj.registerManager(poolManager, SceneToPool::class.java)

        val actual = testObj.nuw(SceneToPool::class.java)

        Assert.assertEquals(actual.someValue.toLong(), poolManager.initialise_value.toLong())
    }

    @Test
    fun getNew_NotPoolable_AfterRecycling() {
        val poolManager = ScenePoolManager(3, 5)
        testObj.registerManager(poolManager, SceneToPool::class.java)

        var actual = testObj.nuw(SceneToPool::class.java)

        Assert.assertEquals(actual.someValue.toLong(), poolManager.initialise_value.toLong())

        testObj.recycle(actual)

        actual = testObj.nuw(SceneToPool::class.java)

        Assert.assertEquals(actual.someValue.toLong(), poolManager.reset_value.toLong())
    }

    @Test
    fun getNew_Poolable_AfterRecycling() {

        val actual = testObj.nuw(PoolableScene::class.java)


        testObj.recycle(actual)

        val actual_v2 = testObj.nuw(PoolableScene::class.java)

        Assert.assertSame(actual, actual_v2)
        Assert.assertTrue(actual.reset)

    }

    class ScenePoolManager(internal var initialise_value: Int, internal var reset_value: Int) : PoolManager<SceneToPool> {

        override fun reset(sceneToPool: SceneToPool) {
            sceneToPool.someValue = reset_value
        }

        override fun create_new(): SceneToPool {
            val newObj = SceneToPool()
            newObj.someValue = initialise_value
            return newObj
        }

        override fun dispose(sceneToPool: SceneToPool) {

        }
    }

    class SceneToPool {
        internal var someValue: Int = 0
    }

    class PoolableScene : Poolable {

        internal var reset = false

        override fun reset() {
            reset = true
        }

    }

}

